client/app/components/dashboards/add-widget-dialog.js

```diff
@@ -58,74 +58,45 @@ const AddWidgetDialog = {
-
-      // Find first free row for each column
-      const bottomLine = _.chain(this.dashboard.widgets)
-        .map((w) => {
-          const options = _.extend({}, w.options);
-          const position = _.extend({ row: 0, sizeY: 0 }, options.position);
-          return {
-            left: position.col,
-            top: position.row,
-            right: position.col + position.sizeX,
-            bottom: position.row + position.sizeY,
-            width: position.sizeX,
-            height: position.sizeY,
-          };
-        })
-        .reduce((result, item) => {
-          const from = Math.max(item.left, 0);
-          const to = Math.min(item.right, result.length + 1);
-          for (let i = from; i < to; i += 1) {
-            result[i] = Math.max(result[i], item.bottom);
-          }
-          return result;
-        }, _.map(new Array(dashboardGridOptions.columns), _.constant(0)))
-        .value();
-
-      // Go through columns, pick them by count necessary to hold new block,
-      // and calculate bottom-most free row per group.
-      // Choose group with the top-most free row (comparing to other groups)
-      const position = _.chain(_.range(0, dashboardGridOptions.columns - width + 1))
-        .map(col => ({
-          col,
-          row: _.chain(bottomLine)
-            .slice(col, col + width)
-            .max()
-            .value(),
-        }))
-        .sortBy('row')
-        .first()
-        .value();
-
       const widget = new Widget({
         visualization_id: this.selectedVis && this.selectedVis.id,
         dashboard_id: this.dashboard.id,
         options: {
           isHidden: this.isTextBox() && this.isHidden,
-          position: {
-            // Place new widget below all others
-            col: position.col,
-            row: position.row,
-            sizeX: width,
-            // Auto-height by default
-            sizeY: -1,
-          },
+          position: {},
         },
         text: this.text,
       });
 
-      widget.$save().then((response) => {
-        // update dashboard layout
-        this.dashboard.version = response.version;
-        this.dashboard.widgets.push(new Widget(response.widget));
-        this.close();
-      }).catch(() => {
-        toastr.error('Widget can not be added');
-      }).finally(() => {
-        this.saveInProgress = false;
-      });
+      widget.$save()
+        .then((response) => {
+          // update dashboard layout
+          this.dashboard.version = response.version;
+
+          response.widget.options = _.extend(
+            {},
+            response.widget.options,
+            { position: {} },
+          );
+          const w = new Widget(response.widget);
+
+          const position = this.dashboard.calculateNewWidgetPosition(w);
+          w.options.position.col = position.col;
+          w.options.position.row = position.row;
+
+          // Save it with new position
+          return w.$save().then(() => w);
```

Comment:
Why not calculate the position before saving, thus eliminating the need to call the server twice?

Reply:
Because before the first save, `visualization` property is not available (and, as i understand, only server can populate it), but it's needed to properly calculate dimensions of widget.

Reply:
If I'm not mistaken, `this.selectedVis` is the same object the server will return.

Reply:
I'll check, and if they are the same - I'll remove second request.